appcelerator/titanium_mobile
# chore(android): improve gradle build fault tolerance

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-27718
- https://jira.appcelerator.org/browse/TIMOB-27720

**Summary:**
- [TIMOB-27718](https://jira.appcelerator.org/browse/TIMOB-27718) Log build warnings if "res" files have invalid names.
  * Restores old Titanium behavior where "res" file names are automatically renamed if invalid.
  * Will log build warnings if invalid along with a suggestion on what to rename it to.
- [TIMOB-27720](https://jira.appcelerator.org/browse/TIMOB-27720) Titanium SDK build should favor "python2" tool if available.
  * Avoids SDK build failures on systems that have python 3 installed.
  * Note: We plan on removing our python usage in the future.
- Now setting Java/Kotlin compiler max heap size via gradle.
  * Avoids `OutOfMemory` exception when building very large projects.
  * Uses existing `android.javac.maxmemory` setting.
  * Uses existing `android.dx.maxmemory` setting.
- Improved "AndroidManifest.xml" fault tolerance for app builds.
  * Now filters out `xmlns:android` namespaces attributes from child XML elements. Can only be defined in root `<manifest/>` element. CLI used to wrongly inject them.

---
**Test:**
1. Create a Classic Titanium app project.
2. Unzip "Resources.zip" attached to [TIMOB-27718](https://jira.appcelerator.org/browse/TIMOB-27718) to project's "Resources" folder.
3. Copy below "tiapp.xml" settings to project.
4. Copy below "app.js" to project.
5. Build and run on Android.
6. Verify that you see 2 Appc logos and a gray button image with dark border onscreen.
7. Verify that you see the following build warnings in the log.
```
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/1-appc-dev-logo.jpg
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Names cannot start with a number.
[WARN]  - Titanium will rename to: _1_appc_dev_logo.jpg
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/ButtonHoloDark.9.png
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Titanium will rename to: buttonholodark.9.png
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/Logos/Appc-Transparent.png
[WARN]  - Files cannot be put into subdirectories.
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Titanium will rename to: logos_appc_transparent.png
```

tiapp.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<uses-permission xmlns:android="" android:name="android.permission.FOREGROUND_SERVICE"/>
			<uses-permission xmlns:android="http://schemas.android.com/apk/res/android" android:name="android.permission.WAKE_LOCK"/>
			<application xmlns:android="">
				<meta-data xmlns:android="" android:name="my.metadata" android:value="Hello World"/>
			</application>
		</manifest>
	</android>
</ti:app>
```

app.js
```javascript
var imagePath;
var window = Ti.UI.createWindow();
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	scrollType: "vertical",
	showHorizontalScrollIndicator: false,
	shorVerticalScrollIndicator: true,
	backgroundColor: "gray",
});
{
	imagePath = "/Resources/images/1-appc-dev-logo.jpg";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createImageView({ image: imagePath }));
}
{
	imagePath = "/Resources/images/Logos/Appc-Transparent.png";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createImageView({ image: imagePath }));
}
{
	imagePath = "/Resources/images/ButtonHoloDark.9.png";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createView({ backgroundImage: imagePath, width: "50%", height: "25%" }));
}
window.add(scrollView);
window.open();
```
